Documentation/ceph-object-bucket-notifications.md

```diff
@@ -0,0 +1,169 @@
+
+## Topics
+A CephBucketTopic represents an endpoint (of types: Kafka, AMQP0.9.1 or HTTP), or a specific resource inside this endpoint (e.g a Kafka or an AMQP topic, or a specific URI in an HTTP server). The CephBucketTopic also hold any additional info needed for the RADOS Gateway (RGW) to connect to that endpoint. Topics don't belong to a specific bucket or notification. Notifications from multiple buckets may be sent on the same topic, and one bucket (via multiple CephBucketNotifications) may send notifications on multiple topics.
+
+## Notifications
+A CephBucketNotification define: topic, events and filters of a bucket notifications. Meaning, it defines what triggers the notification, and where to send the notifications to. However, to complete the definition of the bucket notification, it must be associated with an OBC. This is done outside of the CephBucketNotification resource, via adding labels to an OBC with the following format:
+```yaml
+bucket-notification-<notification name>: <notification name>
+```
+> Note that the CephBucketTopic, CephBucketNotification and OBC must all belong to the same namespace
+
+## Notification Reliability and Delivery
+Notifications may be sent synchronously, as part of the operation that triggered them. In this mode, the operation is acked only after the notification is sent to the topic’s configured endpoint, which means that the round trip time of the notification is added to the latency of the operation itself.
+> Note that the original triggering operation will still be considered as successful even if the notification fail with an error, cannot be delivered or times out
+
+Notifications may also be sent asynchronously. They will be committed into persistent storage and then asynchronously sent to the topic’s configured endpoint. In this case, the only latency added to the original operation is of committing the notification to persistent storage.
+> Note that if the notification fail with an error, cannot be delivered or times out, it will be retried until successfully acked
+
+## Sample
+
+### CephBucketTopic Custom Resource
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: CephBucketTopic
+metadata:
+  name: my-topic [1]
+  namespace: my-app-space [2]
+spec:
+   # use one of the following endpoints
+  endpoint: http://my-notification-endpoint:8080 [3]
+  # endpoint: http://my-notification-endpoint:8080/my-topic
+  # endpoint: https://my-notification-endpoint:8443
+  # endpoint: amqp://my-rabbitmq-service:5672/vhost1
+  # endpoint: amqps://my-rabbitmq-service:5671/vhost1
+  # endpoint: kafka://my-kafka-service:9092
+  objectStoreName: my-store [4]
+  objectStoreNamespace: rook-ceph [5]
+  opaqueData: my@email.com [6]
+  persistent: false [7]
+  http: [8]
+    verifySSL: false [9]
+  # amqp: [10]
+  #   verifySSL: false [11]
+  #   ackLevel: broker [12]
+  #   exchange: my-exchange [13]
+  # kafka: [14]
+  #   verifySSL: false [15]
+  #   ackLevel: broker [16]
+  #   useSSL: false [17]
+
```

Comment:
this makes sense.
but, since this is a change to the API, we should probably do that as part of this PR (to prevent compatibility and migration issues)

Reply:
Maybe `protocol` could be changed to `endpoint` here. I think that is more descriptive that it defines where the notifications will "end" up.

Reply:
will change:
protocol --> endpoint
endpoint --> url